WordPress/openverse
# Add alter data step to the data refresh DAG

## Fixes



Fixes #4541 by @AetherUnbound

## Description



This PR adds a data filtering step (named `alter_data`) to the data refresh DAG. These steps are largely based on those of the `cleanup.py` file from the ingestion server.

However, unlike what's stated in #4541, the code is **not** copied over directly. I initially tried to do this to reduce the amount of work necessary to set this step up, but I ran into https://github.com/apache/airflow/issues/14896 and was unable to leverage the `multiprocessing` library within the context of a single Airflow task. I opted to take an approach similar to #4572 (indeed, producing a lot of similar code!) and instead use Airflow's mapped tasks to split the data across a number of workers. This actually simplified the code _significantly_ (along with removing the other cleanup steps and the generic field handling). Turns out: using Airflow for the thing Airflow was intended for makes things easier! Yay!

All that to say, this adds a task group to the data refresh after the `copy_data` step and before the distributed reindexing which filters out any tags that meet the following criteria:
- They're malformed
- They match or are included in our denylists
- They are below a certain accuracy
- They are from an unvetted provider

> [!IMPORTANT]
> This is a divergence from current ingestion server behavior because we will **also be filtering data for the `audio` table now**. Previous filtering was done only on the image table, so now all our data refresh DAGs will receive the same filtering (which is good!).

![image](https://github.com/user-attachments/assets/5e6ab0e7-e15a-4f6d-8656-fa070f4aeeda)


## Testing Instructions


> [!WARNING]
> A new variable, `AIRFLOW_VAR_DATA_REFRESH_ALTER_BATCH_SIZE=1000`, has been added to the `env.template` file. Be sure to add this to your own `catalog/.env` file before testing!

1. Run `ov just c && ov just i` to bring the catalog stack up (along with the downstream API database)
2. Enable the `production_audio_data_refresh` DAG (this should kick off a run automatically)
3. Observe that the `alter_table_data` block runs and succeeds
4. If you've set the above variable appropriately, you should see **5** mapped tasks for `alter_data_batch`
5. Click on mapped index **3** and view the logs, you should see `1` returned as the XCom and a note in the logs that 1 row was affected
6. (Optional) Remove the aforementioned Variable from your environment and run steps 1-5 again. See that only 1 mapped task is created, and the default batch size (100,000) is used instead of 1,000

Tests should also pass!


## Checklist


      PRs) or the media properties generator (`ov just catalog/generate-docs media-props`
      for the catalog or `ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
